weedcoco/validation.py

```diff
@@ -20,6 +19,18 @@ class ValidationError(Exception):
     pass
 
 
+class JsonValidationError(ValidationError):
+    def __init__(self, message, jsonschema_errors=None):
+        super().__init__(message)
+        self.jsonschema_errors = jsonschema_errors
+
+    def error_path_message(self):
+        return [
+            (f"Error in {' / '.join(str(p) for p in error.path)}", error.message)
```

Comment:
Why build strings when we want this to be transmitted to the client side as a rich JSON blob in any case?

Reply:
My thinking of design is that the error handling in the frontend only takes error message from BadRequest. If we want to send the wholes list of errors, might end up having an overweight exception discussion in the front.

Reply:
I don't think there's anything wrong with passing a kb of error messages back to the frontend, as long as the frontend limits/paginates the output. Or the backend can limit the total number of errors reported.

```
{
 "error_type": "jsonschema",
 "n_errors_found": "100",
 "error_details": [  # lists first 10 errors, say
    "path": ["categories", 0, "name"],
    "message": ...,
    "schema": ...,
 ]
}
```


Another thought: if the number of errors is really large, then it's likely that the number of _distinct_ errors is much smaller. As such you could instead group the content by message/schema:
```
{
 "error_type": "jsonschema",
 "n_errors_found": "100",
 "error_details": [
    "instances": [
      {"path": ["categories", 0, "name"], "value": "weed 1"},
      {"path": ["categories", 1, "name"], "value": "weed 2"},
      {"path": ["categories", 2, "name"], "value": "weed 3"}
    ],
    "message": ...,
    "schema": ...,
 ]
}
```

Reply:
Hmm... that doesn't work because the value is included in the error message.

But you could group failures by the `schema_path` or by the path without any array indices, like `["categories", "name"]`.

So the UI could say:
```
Total 11 errors

10 errors in the `name` field of categories

Showing first 3:
* 'weed 1' does not match ...
* 'weed 2' does not match ...
* 'weed 3' does not match ...

About the `name` field:
          A category for an annotation, consisting of an agricultural role and
          optionally a biological name, separated by ": ". Currently allowed
          agricultural roles are "weed" and "crop".  The "none" category may be
          used to indicate that no weed or crop is identified in the image.
          Annotations should be labelled with the most specific category possible,
          usually "weed: <species name>", but super categories may be more general.
          Biological names should be lowercase.
          For an unspecified species of weed, "weed: UNSPECIFIED" may be used.
          For example, to label an oat crop in an image, the category name would be
          "crop: avena sativa". Its super categories may include "crop: grasses"
          and "crop". Oats as a volunteer crop would be "weed: avena sativa".

1 error in the `bbch_growth_range` field of agcontexts:
* ....


About the `bbch_growth_range` field:
      BBCH Growth Stage range
      A pair of BBCH Growth Stage codes (00-99) representing the minimum and
      maximum crop growth stage in the images.
      If this AgContext is not in a cropping environment, or is in fallow, use value "na".
```

Reply:
But basically, for now, return all the errors in a structured form, and the client side can digest them for display, I think.